### PR TITLE
feat(anvil): Add `--fork-chain-id` to enable offline-start mode

### PR DESCRIPTION
**Pull request by xphoniex**
Signed-off-by: xphoniex <xphoniex@users.noreply.github.com>

## Motivation

https://github.com/foundry-rs/foundry/issues/3837:  idea is being able to run `anvil` using a local cache file, without having to contact remote endpoint as long as the required state is in the local cache.

## Solution

When starting `anvil` from a fork, it first retrieves over rpc the `chain_id` of remote, and then creates the `block` and compares that with the one from local cache file, if they are the same, it'll populate in-memory db with local cache. 

Since we want to avoid unnecessary contact with remote, we'll have to pass `--fork-chain-id` so that we can know which local cache file to read, and since we don't know the state of `block`, we can either skip the check (as we're comparing `default` block vs. the one in the cache file) or we can load db twice, second load would be after we substitute the values for `block` with the ones we received from local cache file. I've opted for the former here.

Once `anvil` is up and running, rest of the operation would run just the same, and if any state data is missing it'll be retrieved from remote, and on shutdown, flushed to disk.

### CODE REVIEW

File: anvil/src/cmd.rs [hunk 1]
```diff
@@ -364,6 +365,18 @@ pub struct AnvilEvmArgs {
     #[clap(long, requires = "fork_url", value_name = "BACKOFF", help_heading = "FORK CONFIG")]
     pub fork_retry_backoff: Option<u64>,
 
+    /// Give precedence to local cache file over fetching state from a remote endpoint.
+    ///
+    /// You still must pass both `--fork-url` and `--fork-block-number`, and already have your required
+    /// state cached on disk, anything missing locally would be fetched from the remote.
```
Comment (mattsse) [inline anvil/src/cmd.rs:371]: Can we add a smol note here that this is mainly intended for offline mode?

File: anvil/src/cmd.rs [hunk 2]
```diff
@@ -364,6 +365,18 @@ pub struct AnvilEvmArgs {
...
+    #[clap(
+        long,
+        help_heading = "FORK CONFIG",
+        value_name = "CHAIN",
+        requires = "fork_block_number"
+    )]
+    pub fork_chain_id: Option<ethers::prelude::U256>,
```
Comment (mattsse) [inline anvil/src/cmd.rs:378]: let's use the `Chain` type here, ref `--chain` argument

File: anvil/src/config.rs [hunk 3]
```diff
@@ -111,6 +111,8 @@ pub struct NodeConfig {
     pub eth_rpc_url: Option<String>,
     /// pins the block number for the state fork
     pub fork_block_number: Option<u64>,
+    /// specifies the chain id for cache file
```
Comment (mattsse) [inline anvil/src/config.rs:114]: please duplicate offline docs here.

File: evm/src/executor/fork/cache.rs [hunk 4]
```diff
@@ -32,13 +32,16 @@ impl BlockchainDb {
     ///   - the file the `cache_path` points to, does not exist
     ///   - the file contains malformed data, or if it couldn't be read
     ///   - the provided `meta` differs from [BlockchainDbMeta] that's stored on disk
-    pub fn new(meta: BlockchainDbMeta, cache_path: Option<PathBuf>) -> Self {
+    pub fn new(meta: BlockchainDbMeta, cache_path: Option<PathBuf>, skip_check: bool) -> Self {
```
Comment (mattsse) [inline evm/src/executor/fork/cache.rs:35]: this is for the "unrelated" metadata issue you mentioned, right?

I'd rather keep the new function and another like `fn with_cache_check(.., bool)` or similar and then new calls this with false

**Review by mattsse — CHANGES_REQUESTED** [review]

File: anvil/src/config.rs (hunk 3 above)
Comment (xphoniex) [reply anvil/src/config.rs]: I just mentioned offline-start [here](https://github.com/foundry-rs/foundry/compare/3f9ba887901333c50eed57ac4078b3f946fda521..fc12efd78d93ada8530600c2a393418e581fc8bd#diff-3002b4d2e85259099616d3b61841c0113e6bb7de324d701cf2467148cd14bb90R114). You don't want the whole thing, do you?

**Review by mattsse — APPROVED** [review]
thanks!

lgtm

**Pull request merged**